cmd/zc_traverser_blob_versions.go

```diff
@@ -0,0 +1,121 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/azure-pipeline-go/pipeline"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+type blobVersionsTraverser struct {
+	rawURL                      *url.URL
+	p                           pipeline.Pipeline
+	ctx                         context.Context
+	recursive                   bool
+	includeDirectoryStubs       bool
+	incrementEnumerationCounter enumerationCounterFunc
+	listOfVersionIds            chan string
+}
+
+func (t *blobVersionsTraverser) isDirectory(isSource bool) bool {
+	isDirDirect := copyHandlerUtil{}.urlIsContainerOrVirtualDirectory(t.rawURL)
+
+	// Skip the single blob check if we're checking a destination.
+	// This is an individual exception for blob because blob supports virtual directories and blobs sharing the same name.
+	if isDirDirect || !isSource {
+		return isDirDirect
+	}
+
+	_, err := t.getBlobProperties("")
+
+	return err != nil
+}
+
+func (t *blobVersionsTraverser) getBlobProperties(versionID string) (props *azblob.BlobGetPropertiesResponse, err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+	blobURLParts.BlobName = strings.TrimSuffix(blobURLParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)
+	if versionID != "" {
+		blobURLParts.VersionID = versionID
+	}
+
+	blobURL := azblob.NewBlobURL(blobURLParts.URL(), t.p)
+	props, err = blobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{})
+	return props, err
+}
+
+func (t *blobVersionsTraverser) traverse(preprocessor objectMorpher, processor objectProcessor, filters []objectFilter) (err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+
+	versionID, ok := <-t.listOfVersionIds
+	for ; ok; versionID, ok = <-t.listOfVersionIds  {
+		blobProperties, err := t.getBlobProperties(versionID)
+
+		if err != nil {
+			return err
+		}
+
+		if blobProperties == nil {
+			panic("isBlob should never be set if getting properties is an error")
+		}
+		blobURLParts.VersionID = versionID
+		storedObject := newStoredObject(
+			preprocessor,
+			getObjectNameOnly(strings.TrimSuffix(blobURLParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)),
+			"",
+			common.EEntityType.File(),
+			blobProperties.LastModified(),
+			blobProperties.ContentLength(),
+			blobProperties,
+			blobPropertiesResponseAdapter{blobProperties},
+			common.FromAzBlobMetadataToCommonMetadata(blobProperties.NewMetadata()),
+			blobURLParts.ContainerName,
+		)
+
+		storedObject.Metadata["versionID"] = versionID
```

Comment:
>storedObject.Metadata["versionID"] = versionID [](start = 2, length = 46)

this works but it's a hacky solution. Metadata has a very specific meaning in the context of blob service; it is a set of key/value pairs specified by the user; we should use a dedicated field to store the version ID. If we sneak in the version ID this way, it'll be hard for the future devs to notice this and we might accidentally overwrite the metadata when blob -> blob copy is supported for versions.

Reply:
Yes that's entirely correct. In fact Narasimha raised the same concern. I'm also thinking of a unique field but there is always a chance that user may use that dedicated field to set metadata.

Reply:
Done